File: android/app/src/main/java/bd/sicip/qavisit/pdf/QaReportHtml.kt
// QA Monitoring Visit Report (Annex-3) printable HTML -- QA report spec §7. A separate layout
// from pdf/ReportHtml.kt's surprise-visit report: this one follows the paper Annex-3 form
// heading-by-heading and table-by-table, adding exactly one column (REMARKS) to every criteria
// table. Mirrors web/src/lib/qareporthtml.js (the other agent's own port of this same spec
// section) -- kept in lockstep the same way pdf/ReportHtml.kt and reporthtml.js are: same CSS,
// same structure, hand-kept copies since there is no shared module either platform can import.
//
// pure string building, no android imports -- unit tested by test/.../pdf/QaReportHtmlTest.kt.
// The actual print-to-PDF goes through pdf/BillPrinter.kt's WebView pipeline, same as every other
// report PDF in this app (ui/reports/ReportPdf.kt's shareReportPdf picks this builder over
// pdf/ReportHtml.kt's when template.id == "qa").
package bd.sicip.qavisit.pdf

import bd.sicip.qavisit.domain.report.CriteriaItem
import bd.sicip.qavisit.domain.report.Field
import bd.sicip.qavisit.domain.report.ReportBlock
import bd.sicip.qavisit.domain.report.ReportData
import bd.sicip.qavisit.domain.report.ReportSection
import bd.sicip.qavisit.domain.report.ReportTemplate
import bd.sicip.qavisit.domain.report.normalize
import bd.sicip.qavisit.domain.report.printedRemarks
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonPrimitive

private fun esc(s: String?): String = (s ?: "")
    .replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;").replace("\"", "&quot;")

private fun blank(v: String?): Boolean = v == null || v.trim().isEmpty()

private fun escMultiline(s: String): String = esc(s).replace("\n", "<br>")

private fun JsonObject.stringOrNull(key: String): String? = this[key]?.jsonPrimitive?.contentOrNull

// "from dd/mm/yyyy to dd/mm/yyyy" out of the visit's own start/end (yyyy-mm-dd, this app's
// storage format) -- Annex-3's exact wording (spec §7's header line example).
private fun ddmmyyyy(isoDate: String): String {
    val parts = isoDate.split("-")
    if (parts.size != 3) return isoDate
    return "${parts[2]}/${parts[1]}/${parts[0]}"
}

// print CSS -- A4 portrait, Times New Roman 11pt, black on white, thin borders (spec §7), page
// numbers via @page counters, footer text fixed per spec ("SICIP Quality Assurance Visit Report").
private val CSS = """
  @page {
    size: A4 portrait;
    margin: 10mm 11mm 12mm;
    @bottom-left { content: "SICIP Quality Assurance Visit Report"; font: 8pt "Times New Roman", Times, serif; color: #333; }
    @bottom-right { content: "Page " counter(page) " of " counter(pages); font: 8pt "Times New Roman", Times, serif; color: #333; }
  }
  * { box-sizing: border-box; }
  body { margin: 0; font-family: "Times New Roman", Times, serif; font-size: 11pt; line-height: 1.3; color: #000; background: #fff; }

  .annex { text-align: right; font-weight: 700; }
  .program { text-align: center; font-weight: 700; margin-top: 2pt; }
  .title { text-align: center; font-weight: 700; font-size: 13pt; margin: 2pt 0 8pt; }

  .kv { margin: 0 0 8pt; }
  .kv div { margin: 1pt 0; }
  .kv b { font-weight: 700; }

  h2.sh { font-weight: 700; text-transform: uppercase; font-size: 11.5pt; margin: 10pt 0 2pt; }
  p.intro { font-style: italic; margin: 0 0 4pt; }
  h3 { font-weight: 700; font-size: 11pt; margin: 6pt 0 2pt; }

  table { width: 100%; border-collapse: collapse; table-layout: fixed; margin-bottom: 6pt; }
  th, td { border: 0.75pt solid #000; padding: 3pt 4pt; vertical-align: top; font-size: 10.5pt; }
  th { font-weight: 700; text-align: center; }
  td.sl { text-align: center; }
  td.heading-row { font-weight: 700; }

  ul.remarks { margin: 0; padding-left: 14pt; }
  ul.remarks li { margin: 0 0 2pt; }

  .field-line { margin: 1pt 0; }
  .field-line b { font-weight: 700; }
  .field-box { margin: 2pt 0 6pt; }
  .field-box .lbl { font-weight: 700; }
  .field-box .box { border: 0.75pt solid #000; min-height: 16pt; padding: 2pt 4pt; margin-top: 1pt; }

  ul.points { margin: 2pt 0; padding-left: 16pt; }

  .sign { display: flex; justify-content: space-between; margin-top: 16pt; }
  .sign .officers div { margin: 2pt 0; }
""".trimIndent()

// header block: top-right "Annex-3", centred bold program line, centred bold title, then the
// fixed key/value lines exactly as Annex-3 prints them (spec §7).
private fun headerHtml(template: ReportTemplate, data: ReportData): String {
    val annex = template.annex ?: "Annex-3"
    val dateFrom = data.field("date_from")
    val dateTo = data.field("date_to")
    val dates = if (!blank(dateFrom) && !blank(dateTo)) {
        "from ${ddmmyyyy(dateFrom)} to ${ddmmyyyy(dateTo)}"
    } else {
        ""
    }
    return """
      <div class="annex">${esc(annex)}</div>
      <div class="program">${esc(template.program)}</div>
      <div class="title">${esc(template.title)}</div>
      <div class="kv">
        <div><b>Name of Training TI/TC Visited :</b> ${esc(data.field("ti_name"))}</div>
        <div><b>Name of the Association/Provider :</b> ${esc(data.field("provider"))}</div>
        <div><b>Date(s) of Visit :</b> ${esc(dates)}</div>
        <div><b>Name(s) &amp; Designation(s) of Visiting Officer(s) :</b> ${esc(data.field("officers"))}</div>
      </div>
    """.trimIndent()
}

// one criteria table (spec §7): Sl. | QUALITY CRITERIA | EVIDENCE | REMARKS, widths ~7/31/31/31%.
// a heading item (no options) is one row with its Sl + text spanning the other three columns; a
// real item's REMARKS cell is domain/report/Remarks.kt's printedRemarks as a bulleted list, blank
// when there's nothing to print (never a placeholder).
private fun criteriaBlockHtml(block: ReportBlock.Criteria, data: ReportData): String {
    val rows = block.items.joinToString("") { item ->
        if (item.heading) {
            "<tr><td class=\"sl\">${esc(item.no)}</td><td class=\"heading-row\" colspan=\"3\">${esc(item.text)}</td></tr>"
        } else {
            val evidenceHtml = escMultiline(item.evidence ?: "")
            val remarksHtml = remarksListHtml(item, data)
            "<tr><td class=\"sl\">${esc(item.no)}</td><td>${esc(item.text)}</td><td>$evidenceHtml</td><td>$remarksHtml</td></tr>"
        }
    }
    val intro = block.intro?.takeIf { it.isNotBlank() }?.let { "<p class=\"intro\">(${esc(it)})</p>" } ?: ""
    return intro +
        "<table><colgroup><col style=\"width:7%\"><col style=\"width:31%\"><col style=\"width:31%\"><col style=\"width:31%\"></colgroup>" +
        "<thead><tr><th>Sl.</th><th>QUALITY CRITERIA</th><th>EVIDENCE</th><th>REMARKS</th></tr></thead>" +
        "<tbody>$rows</tbody></table>"
}

private fun remarksListHtml(item: CriteriaItem, data: ReportData): String {
    val bullets = printedRemarks(item, data)
    if (bullets.isEmpty()) return ""
    return "<ul class=\"remarks\">${bullets.joinToString("") { "<li>${esc(it)}</li>" }}</ul>"
}

// section 1's fields/cards, and the conclusions sections (11-13's cards/fields, 14/15's
// longtext) -- a generic best-effort renderer (label: value lines, longtext as a bulleted list of
// its own non-blank lines, cards as one small table per block) rather than Annex-3's exact
// hand-tuned sub-tables (1.10-1.62's own column layouts, 13's 9 fixed component rows): those
// depend on the real qa-v1.json's exact field/card keys, which don't exist yet at the time this
// was written (see this agent's own build note in DECISIONS.md). Generic rendering still prints
// every value the officer entered, in template order, losing only the paper form's exact spacing
// -- reasonable given the template it must match isn't authored yet.
private fun fieldsBlockHtml(block: ReportBlock.Fields, data: ReportData): String {
    val heading = block.heading?.let { "<h3>${esc(it)}</h3>" } ?: ""
    return heading + block.fields.joinToString("") { field -> fieldHtml(field) { data.field(it) } }
}

private fun fieldHtml(field: Field, value: (String) -> String): String {
    val raw = value(field.key)
    if (field.kind == "longtext") {
        if (blank(raw)) return "<div class=\"field-box\"><div class=\"lbl\">${esc(field.label)}</div><div class=\"box\"></div></div>"
        // "one point per line" fields (spec §2's findings/recommendations/dropout-reasons/steps)
        // print as a bulleted list when they have more than one non-blank line, a plain box
        // otherwise -- either way every line the officer wrote prints, nothing collapsed away.
        val lines = raw.split("\n").map { it.trim() }.filter { it.isNotBlank() }
        val body = if (lines.size > 1) "<ul class=\"points\">${lines.joinToString("") { "<li>${esc(it)}</li>" }}</ul>" else escMultiline(raw)
        return "<div class=\"field-box\"><div class=\"lbl\">${esc(field.label)}</div><div class=\"box\">$body</div></div>"
    }
    return "<div class=\"field-line\"><b>${esc(field.label)} :</b> ${esc(raw)}</div>"
}

private fun cardsBlockHtml(block: ReportBlock.Cards, data: ReportData): String {
    val cards = data.cards(block.key)
    if (cards.isEmpty()) return ""
    val headerCells = block.fields.joinToString("") { "<th>${esc(it.label)}</th>" }
    val rows = cards.joinToString("") { card ->
        val cells = block.fields.joinToString("") { f -> "<td>${esc(card.stringOrNull(f.key))}</td>" }
        "<tr>$cells</tr>"
    }
    return "<table><thead><tr>$headerCells</tr></thead><tbody>$rows</tbody></table>"
}

private fun blockHtml(block: ReportBlock, data: ReportData): String = when (block) {
    is ReportBlock.Criteria -> criteriaBlockHtml(block, data)
    is ReportBlock.Fields -> fieldsBlockHtml(block, data)
    is ReportBlock.Cards -> cardsBlockHtml(block, data)
    // qa-v1.json has neither of these block types (spec §2) -- nothing to print if one ever
    // sneaks in, rather than guessing at a layout for it.
    is ReportBlock.Checklist -> ""
    is ReportBlock.Flags -> ""
}

private fun sectionHtml(section: ReportSection, data: ReportData): String {
    val heading = "<h2 class=\"sh\">${esc(section.badge)}. ${esc(section.title.uppercase())}</h2>"
    val blocks = section.blocks.joinToString("") { blockHtml(it, data) }
    return heading + blocks
}

// spec §7's signature block: one row per officer named in the header's "officers" field
// (newline- or semicolon-separated, same free text the header line already prints), left column
// names+designations, right column a dotted signature line each.
private fun signatureHtml(data: ReportData): String {
    val officers = data.field("officers").split(Regex("[\n;]")).map { it.trim() }.filter { it.isNotBlank() }
    if (officers.isEmpty()) return ""
    val names = officers.mapIndexed { i, name -> "<div>${i + 1}) ${esc(name)}</div>" }.joinToString("")
    val lines = officers.joinToString("") { "<div>&hellip;&hellip;&hellip;&hellip;&hellip;&hellip;&hellip;&hellip;&hellip;&hellip;&hellip;</div>" }
    return """
      <div class="sign">
        <div class="officers"><b>Name (s) of the Visiting Officer(s) with designation</b>$names</div>
        <div style="text-align:right"><b>Signatures</b>$lines</div>
      </div>
    """.trimIndent()
}

// pure fn: template + report data -> full print HTML, Annex-3 layout (spec §7). Only ever called
// for template.id == "qa" (ui/reports/ReportPdf.kt's shareReportPdf branches on that); a surprise
// report never reaches this file.
fun buildQaReportHtml(template: ReportTemplate, data: ReportData): String {
    val normalizedData = normalize(template, data)
    val sections = template.sections.joinToString("") { sectionHtml(it, normalizedData) }
    return "<!doctype html><html><head><meta charset=\"utf-8\"><title>${esc(template.title)}</title>" +
        "<style>$CSS</style></head><body>" +
        headerHtml(template, normalizedData) + sections + signatureHtml(normalizedData) +
        "</body></html>"
}

File: web/src/lib/qareporthtml.js
// filled QA visit report (Annex-3 exact layout + a 4th Remarks column) as printable HTML --
// spec docs/superpowers/plans/2026-09-25-qa-report.md section 7. A separate layout from the
// surprise report's reporthtml.js: Annex-3's own heading-by-heading, table-by-table shape,
// Times New Roman, "Annex-3" top right, program + title centred. Caller opens the html in a new
// window + window.print(), same pattern as reporthtml.js/billhtml.js.
import { buildRemarks, printedRemarks } from './remarks.js'
import * as reportTemplateModule from './reporttemplate.js'
import { feedbackGrid } from './feedbackgrid.js'

function esc(s) {
  return String(s ?? '').replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;').replace(/"/g, '&quot;')
}

function blank(v) {
  return v == null || String(v).trim() === ''
}

function escMultiline(s) {
  return esc(s).replace(/\n/g, '<br>')
}

// same defensive normalize()-or-syncLinks-or-passthrough fallback as reporthtml.js -- pure fn,
// never mutates the caller's data.
function withNormalizedData(template, data) {
  const normalize = reportTemplateModule.normalize
  const step = typeof normalize === 'function' ? normalize : null
  if (!step) return data
  const cloned = JSON.parse(JSON.stringify(data))
  return step(template, cloned) || cloned
}

// "2026-09-25" -> "25/09/2026" (Annex-3's own date shape); blank/malformed values pass through.
function ddmmyyyy(value) {
  const match = /^(\d{4})-(\d{2})-(\d{2})/.exec(String(value ?? ''))
  if (!match) return blank(value) ? '' : String(value)
  return `${match[3]}/${match[2]}/${match[1]}`
}

// non-blank lines of a longtext field -> lowercase roman numerals "i) ii) iii) ..." (Annex-3's
// own list style for 1.61/1.62, 14, 15).
const ROMAN = ['i', 'ii', 'iii', 'iv', 'v', 'vi', 'vii', 'viii', 'ix', 'x', 'xi', 'xii', 'xiii', 'xiv', 'xv', 'xvi', 'xvii', 'xviii', 'xix', 'xx']
function romanLines(text) {
  const lines = String(text ?? '').split('\n').map((l) => l.trim()).filter(Boolean)
  return lines.map((line, i) => `<div class="roman-line"><span class="roman">${ROMAN[i] ?? i + 1})</span>${esc(line)}</div>`).join('')
}

function fieldsMap(data) {
  return data.fields || {}
}

function kvLine(label, value) {
  return `<div class="kv-line"><b>${esc(label)} :</b> ${esc(value)}</div>`
}

// the Annex-3 header block that sits ABOVE "1. General Information" -- ti_name/provider/address/
// dates/officers. Everything else in section 1's `fields` block (status/purposes/team) prints
// inside section 1 itself as i)/ii)/iii).
function headerKvHtml(data) {
  const f = fieldsMap(data)
  const dateLine = `from ${ddmmyyyy(f.date_from) || '__/__/____'} to ${ddmmyyyy(f.date_to) || '__/__/____'}`
  return `<div class="kv">
    ${kvLine('Name of Training TI/TC Visited', f.ti_name)}
    ${kvLine('Name of the Association/Provider', f.provider)}
    ${kvLine('Address of Training Institute and Contact', f.address)}
    <div class="kv-line"><b>Date(s) of Visit :</b> ${esc(dateLine)}</div>
    ${kvLine("Name(s) & Designation(s) of Visiting Officer(s)", f.officers)}
  </div>`
}

function cardsTableHtml(fields, entries, minRows) {
  const rows = entries.length >= minRows ? entries : entries.concat(Array.from({ length: minRows - entries.length }, () => ({})))
  const header = `<tr>${fields.map((f) => `<th>${esc(f.label)}</th>`).join('')}</tr>`
  const body = rows.map((entry) => `<tr>${fields.map((f) => `<td>${escMultiline(entry[f.key])}</td>`).join('')}</tr>`).join('')
  return `<table class="grid">${header}${body}</table>`
}

// 1.10 persons met.
function personsHtml(block, data) {
  const entries = (data.cards && data.cards[block.key]) || []
  return `<div class="sub-h">${esc(block.heading)}</div>${cardsTableHtml(block.fields, entries, block.start || 3)}`
}

// 1.20-1.25 + 1.30-1.34 + comments -- plain label:value lines (short text/date/number fields).
function mouFieldsHtml(block, data) {
  const f = fieldsMap(data)
  const lines = block.fields
    .map((field) => {
      if (field.kind === 'longtext') {
        return `<div class="field-box"><div class="fb-label">${esc(field.label)}</div><div class="fb-body">${escMultiline(f[field.key])}</div></div>`
      }
      if (field.kind === 'choice') {
        const opt = (field.options || []).find((o) => o.id === f[field.key])
        return `<div class="kv-line"><b>${esc(field.label)} :</b> ${esc(opt ? opt.label : '')}</div>`
      }
      return `<div class="kv-line"><b>${esc(field.label)} :</b> ${esc(f[field.key])}</div>`
    })
    .join('')
  return `${block.heading ? `<div class="sub-h">${esc(block.heading)}</div>` : ''}${lines}`
}

// 1.40 MoU courses.
function mouCoursesHtml(block, data) {
  const entries = (data.cards && data.cards[block.key]) || []
  return `<div class="sub-h">${esc(block.heading)}</div>${cardsTableHtml(block.fields, entries, block.start || 4)}`
}

// 1.50 cumulative implementation -- T/F sub-header row, "Job Placed with Percentage" and
// "No. of dropouts with Percentage" print "n (p%)" (placed% of certified, dropout% of enrolled).
function pct(n, of) {
  const num = Number(n)
  const denom = Number(of)
  if (!Number.isFinite(num) || !Number.isFinite(denom) || denom <= 0) return ''
  return ` (${Math.round((num / denom) * 100)}%)`
}
function cumulativeHtml(block, data) {
  const entries = (data.cards && data.cards[block.key]) || []
  const rows = entries.length >= (block.start || 2) ? entries : entries.concat(Array.from({ length: (block.start || 2) - entries.length }, () => ({})))
  const body = rows
    .map((c, i) => {
      const placedT = blank(c.placed_t) ? '' : `${c.placed_t}${pct(c.placed_t, c.certified_t)}`
      const placedF = blank(c.placed_f) ? '' : `${c.placed_f}${pct(c.placed_f, c.certified_f)}`
      const dropoutT = blank(c.dropout_t) ? '' : `${c.dropout_t}${pct(c.dropout_t, c.enrolled_t)}`
      const dropoutF = blank(c.dropout_f) ? '' : `${c.dropout_f}${pct(c.dropout_f, c.enrolled_f)}`
      return `<tr><td>${i + 1}.</td><td>${esc(c.course)}</td><td>${esc(c.target)}</td>` +
        `<td>${esc(c.enrolled_t)}</td><td>${esc(c.enrolled_f)}</td>` +
        `<td>${esc(c.certified_t)}</td><td>${esc(c.certified_f)}</td>` +
        `<td>${esc(placedT)}</td><td>${esc(placedF)}</td>` +
        `<td>${esc(dropoutT)}</td><td>${esc(dropoutF)}</td></tr>`
    })
    .join('')
  return `<div class="sub-h">${esc(block.heading)}</div>
    <table class="grid cumulative">
      <tr><th rowspan="2">S.N.</th><th rowspan="2">Course Name</th><th rowspan="2">Target</th>
        <th colspan="2">Enrolled</th><th colspan="2">Certified</th>
        <th colspan="2">Job Placed with Percentage</th><th colspan="2">No. of dropouts with Percentage</th></tr>
      <tr><th>T</th><th>F</th><th>T</th><th>F</th><th>T</th><th>F</th><th>T</th><th>F</th></tr>
      ${body}
    </table>
    <p class="note-line">${esc(block.note || 'T= Total and F = Female')}</p>`
}

// 1.60 current batches.
function batchesHtml(block, data) {
  const entries = (data.cards && data.cards[block.key]) || []
  return `<div class="sub-h">${esc(block.heading)}</div>${cardsTableHtml(block.fields, entries, block.start || 2)}`
}

// section 1's whole body, in Annex-3 order: i)/ii)/iii), 1.10, 1.20-1.25, 1.30-1.34+comments,
// 1.40, 1.50, 1.60, 1.61/1.62.
function sectionOneBodyHtml(section, data) {
  const f = fieldsMap(data)
  const out = []
  out.push(`<div class="kv-line"><b>i) Status of the Training Institute :</b> ${esc(f.status)}</div>`)
  out.push(`<div class="kv-line"><b>ii) Purpose(s) of this Monitoring Visit :</b> ${esc(f.purposes)}</div>`)
  out.push(`<div class="kv-line"><b>iii) Monitoring Team Members with Designations :</b> ${esc(f.team)}</div>`)
  for (const block of section.blocks) {
    if (block.type === 'cards' && block.key === 'persons') out.push(personsHtml(block, data))
    else if (block.type === 'fields' && block.heading === '1.20 Contract/MoU Information') out.push(mouFieldsHtml(block, data))
    else if (block.type === 'fields' && block.heading && block.heading.startsWith('1.30')) out.push(mouFieldsHtml(block, data))
    else if (block.type === 'cards' && block.key === 'mou_courses') out.push(mouCoursesHtml(block, data))
    else if (block.type === 'cards' && block.key === 'cumulative') out.push(cumulativeHtml(block, data))
    else if (block.type === 'cards' && block.key === 'batches') out.push(batchesHtml(block, data))
    else if (block.type === 'fields' && block.fields.some((fl) => fl.key === 'dropout_reasons')) {
      const reasons = block.fields.find((fl) => fl.key === 'dropout_reasons')
      const steps = block.fields.find((fl) => fl.key === 'dropout_steps')
      out.push(`<div class="kv-line"><b>${esc(reasons.label)} :</b></div>${romanLines(f.dropout_reasons)}`)
      out.push(`<div class="kv-line"><b>${esc(steps.label)} :</b></div>${romanLines(f.dropout_steps)}`)
    }
  }
  return out.join('')
}

// ---- sections 2-10 + 8.1: `criteria` blocks -- Sl. | QUALITY CRITERIA | EVIDENCE | REMARKS ----

function criteriaRowHtml(item, data, no) {
  if (item.heading) {
    return `<tr class="heading-row"><td>${esc(item.no)}</td><td colspan="3">${esc(item.text)}</td></tr>`
  }
  const entry = (data.criteria && data.criteria[item.id]) || {}
  const bullets = printedRemarks(item, entry)
  const remarksHtml = bullets.length ? `<ul>${bullets.map((b) => `<li>${esc(b)}</li>`).join('')}</ul>` : ''
  return `<tr><td>${esc(item.no)}</td><td>${esc(item.text)}</td><td>${escMultiline(item.evidence)}</td><td class="rm">${remarksHtml}</td></tr>`
}

function criteriaBlockHtml(block, data) {
  const heading = block.heading ? `<div class="sub-h">${esc(block.heading)}</div>` : ''
  const intro = block.intro ? `<p class="intro">(${esc(block.intro)})</p>` : ''
  const rows = block.items.map((item) => criteriaRowHtml(item, data)).join('')
  return `${heading}${intro}<table class="grid criteria">
    <colgroup><col style="width:7%"><col style="width:31%"><col style="width:31%"><col style="width:31%"></colgroup>
    <tr><th>Sl.</th><th>QUALITY CRITERIA</th><th>EVIDENCE</th><th>REMARKS</th></tr>
    ${rows}
  </table>`
}

// ---- sections 11/12: anonymous feedback, questions down, respondents across ----
function feedbackTableHtml(block, data) {
  const entries = (data.cards && data.cards[block.key]) || []
  const { tables, comments } = feedbackGrid(block, entries)
  const tablesHtml = tables.map((table) => {
    const header = `<tr>${table.headers.map((h) => `<th>${esc(h)}</th>`).join('')}</tr>`
    const body = table.rows.map((row) => `<tr><td>${esc(row[0])}</td>${row.slice(1).map((v) => `<td class="c">${esc(v)}</td>`).join('')}</tr>`).join('')
    return `<table class="grid">${header}${body}</table>`
  }).join('')
  const commentsHtml = comments.length ? `<div class="sub-h">Comments</div><ul>${comments.map((c) => `<li>${esc(c)}</li>`).join('')}</ul>` : ''
  return `${tablesHtml}${commentsHtml}`
}

// ---- section 13: component-wise strengths & weaknesses, one row per template pair ----
function strengthsWeaknessesHtml(block, data) {
  const f = fieldsMap(data)
  const rows = block.pairs.map((pair, i) =>
    `<tr><td>${i + 1}.</td><td>${esc(pair.component)}</td><td>${escMultiline(f[pair.strength])}</td><td>${escMultiline(f[pair.weakness])}</td></tr>`,
  ).join('')
  return `<table class="grid"><tr><th>S.N.</th><th>Component</th><th>Strengths</th><th>Weakness</th></tr>${rows}</table>`
}

// ---- section 16: improvement plan, minimum 3 rows ----
function planTableHtml(block, data) {
  const entries = (data.cards && data.cards[block.key]) || []
  const rows = entries.length >= 3 ? entries : entries.concat(Array.from({ length: 3 - entries.length }, () => ({})))
  const body = rows.map((entry, i) => `<tr><td>${i + 1}.</td>${block.fields.map((field) => `<td>${escMultiline(entry[field.key])}</td>`).join('')}</tr>`).join('')
  return `<table class="grid"><tr><th style="width:5%">S.N.</th>${block.fields.map((field) => `<th>${esc(field.label)}</th>`).join('')}</tr>${body}</table>`
}

// ---- sections 14/15: bullet lists from lines ----
function bulletListHtml(text) {
  const lines = String(text ?? '').split('\n').map((l) => l.trim()).filter(Boolean)
  if (lines.length === 0) return '<ul><li class="empty-li"></li></ul>'
  return `<ul>${lines.map((l) => `<li>${esc(l)}</li>`).join('')}</ul>`
}

// section.title is stored verbatim from Annex-3 (spec section 2) -- some are fully upper-case
// ("QUALITY MANAGEMENT SYSTEM"), others are mixed-case after an em dash ("TRAINING DELIVERY
// SYSTEM – Physical Resources") or a lower-case parenthetical ("(full appraisal)"); never
// re-case it here.
function sectionHtml(section, data) {
  const heading = `<div class="sh">${esc(section.number)}. ${esc(section.title)}</div>`
  let body = ''
  if (section.key === 's1') {
    body = sectionOneBodyHtml(section, data)
  } else if (section.blocks.some((b) => b.type === 'criteria')) {
    // each block prints its own intro (block.intro) / sub-heading (block.heading, e.g. "8.1 ...")
    body = section.blocks.map((b) => criteriaBlockHtml(b, data)).join('')
  } else if (section.key === 's11' || section.key === 's12') {
    body = feedbackTableHtml(section.blocks[0], data)
  } else if (section.key === 's13') {
    body = strengthsWeaknessesHtml(section.blocks[0], data)
  } else if (section.key === 's14') {
    body = bulletListHtml(fieldsMap(data).findings)
  } else if (section.key === 's15') {
    body = bulletListHtml(fieldsMap(data).recommendations)
  } else if (section.key === 's16') {
    body = planTableHtml(section.blocks[0], data)
  }
  return `${heading}${body}`
}

// officer signature block: one row per officer named in the free-text "officers" field
// (one per line, or "; "-separated), minimum one blank row.
function officerLines(text) {
  const raw = String(text ?? '').split(/\n|;/).map((s) => s.trim()).filter(Boolean)
  return raw.length ? raw : ['']
}
function signatureHtml(data) {
  const names = officerLines(fieldsMap(data).officers)
  const left = names.map((n, i) => `<div>${i + 1}) ${esc(n)}</div>`).join('')
  const right = names.map(() => '<div>……………………………</div>').join('')
  return `<div class="sign">
    <div>Name (s) of the Visiting Officer(s) with designation${left}</div>
    <div class="sign-right">Signatures${right}</div>
  </div>`
}

const CSS = `
  @page {
    size: A4 portrait;
    margin: 12mm 14mm 14mm;
    @bottom-left { content: "SICIP Quality Assurance Visit Report"; font: 8pt "Times New Roman", Times, serif; color: #333; }
    @bottom-right { content: "Page " counter(page) " of " counter(pages); font: 8pt "Times New Roman", Times, serif; color: #333; }
  }
  * { box-sizing: border-box; }
  body { margin: 0; font-family: "Times New Roman", Times, serif; font-size: 11pt; line-height: 1.3; color: #000; background: #fff; }
  .annex { text-align: right; font-weight: 700; }
  .t1, .t2 { text-align: center; font-weight: 700; }
  .t2 { margin-bottom: 8pt; }
  .kv { margin: 6pt 0 10pt; }
  .kv-line { margin: 2pt 0; }
  .sh { font-weight: 700; text-transform: uppercase; margin: 12pt 0 2pt; }
  .sub-h { font-weight: 700; margin: 8pt 0 2pt; }
  .intro { font-style: italic; margin: 0 0 4pt; }
  .roman-line { margin: 1pt 0 1pt 12pt; }
  .roman { display: inline-block; min-width: 22pt; }
  table.grid { width: 100%; border-collapse: collapse; margin: 4pt 0 8pt; }
  table.grid th, table.grid td { border: 0.75pt solid #000; padding: 2pt 4pt; vertical-align: top; font-size: 10pt; }
  table.grid th { font-weight: 700; text-align: center; background: #eee; }
  table.grid td.c { text-align: center; width: 13%; }
  table.criteria th { text-align: left; }
  table.criteria .heading-row td { font-weight: 700; background: #f4f4f4; }
  table.criteria td.rm ul { margin: 0; padding-left: 14pt; }
  table.criteria td.rm li { margin: 1pt 0; }
  .note-line { font-size: 9pt; font-style: italic; margin: -4pt 0 8pt; }
  .field-box { margin: 4pt 0; }
  .fb-label { font-weight: 700; }
  .fb-body { border: 0.75pt solid #666; min-height: 14pt; padding: 2pt 4pt; }
  .sign { display: flex; justify-content: space-between; margin-top: 18pt; gap: 20pt; }
  .sign-right { text-align: right; }
  ul { margin: 2pt 0; padding-left: 16pt; }
  .empty-li { list-style: none; }
`

// pure fn: template + report data + {officerName, submittedAt?, status} -> full print HTML.
export function qaReportHtml(template, data, meta) {
  const normalizedData = withNormalizedData(template, data)
  const sections = (template.sections || []).map((s) => sectionHtml(s, normalizedData)).join('')
  return `<!doctype html><html><head><meta charset="utf-8"><title>${esc(template.title)}</title><style>${CSS}</style></head><body>` +
    `<div class="annex">${esc(template.annex || 'Annex-3')}</div>` +
    `<div class="t1">${esc(template.program)}</div>` +
    `<div class="t2">${esc(template.title)}</div>` +
    headerKvHtml(normalizedData) +
    sections +
    signatureHtml(normalizedData) +
    '</body></html>'
}

// unused meta kept for interface parity with reportHtml (officer/status not shown on the
// Annex-3 layout -- the paper form itself has no such fields).
export function openQaReportPrint(template, data, meta) {
  const html = qaReportHtml(template, data, meta)
  const w = window.open('', '_blank')
  if (!w) return false
  w.document.write(html)
  w.document.close()
  w.onload = () => w.print()
  setTimeout(() => w.print(), 300)
  return true
}
